### ISSUE TITLE
helper for filter form

### ISSUE THREAD

**Issue by woylie** [woylie/flop_phoenix#9]
- [x] implement `Phoenix.Form.HTML`
- [x] filter input helper

**Comment by maltoe** [comment]
Hey @woylie 👋 

We currently have a generated filter form running in an app. A coworker of mine is complaining about the complexity and it certainly feels like a lot of the code belongs into flop_phoenix. What's the state of this feature? What are your plans & ideas for it? We'd like to help you out on this one, if you allow.

Currently what we envision / have partially implemented is something like:

```
filter(
  [{"Name", :foo}, ...],  # foo is the field name
  path_helper: ...,
  path_helper_args: ...,
  ul_attrs: [class: "..."],
  li_attrs: [class: "..."]
)
```

Which would roughly generate markup like this:

```
<form method="get">
  <ul>
    <li>
      <label for="foo_filter">Name</label> 
      <input type="hidden" name="filter[][field]" value="foo" />
      <input type="hidden" name="filter[][op]" value="==" />
      <input type="text" id="field_filter" name="filter[][value]" />
    </li>
  </ul>
</form>
```

Further features that we would like:

* Passing in `phx-submit` for LiveView compatibility, i.e. handling filters with a push_patch. Or with [this upcoming LiveView](https://github.com/phoenixframework/phoenix_live_view/pull/1663) feature it might be possible to set the `push_patch` in flop_phoenix
* Ideally the input field would be determined by the field data type by default, but could be customized by passing a field renderer function in the filter specification
* The `op` value could have options controlling whether it is hidden or not, which default is set, and which values are allowed

One caveat we already discovered is that the URL you build from a filtered `Flop` struct is something like `filters[0][field]=...` which is difficult (impossible?) to create with `Phoenix.HTML.Form` helpers (AFAIK they're always targeting a top-level object a la `filters[filters][0][value]`, you can't have a form for a list of changesets. Not so much an issue for the markup generated in flop_phoenix, but if the API had a way to allow the client to specify input renderer functions, it would be nice if they could have the usual `FormData.t(), atom(), keyword()` signature.

Sorry for the confusing collection of ideas, I'm in a bit of a hurry, but wanted to get this out already.

Would love if we could add this to flop_phoenix! Or something similar, of course.

Best,
malte

**Comment by woylie** [comment]
> 👍 Ideally builder functions at every level :) Especially the `using` option should IMO allow custom markup for specialized input fields.

I'm not sure about builder functions at every level. They make sense for the markup around form fields, but at some point it just becomes easier to write your own template functions to get the form layout you need. Sometimes you may want to group certain fields under a single label etc. That's why I think all the lower level helper functions should be exposed. Let's see how far we can get with a builder function for the `filter_input/1` component first.

`:using` should only be used to set the input helper to use. This follows the pattern laid out [here](https://dashbit.co/blog/dynamic-forms-with-phoenix).

> IMO a `filter_form` version would be a bit more consistent with the pagination and table functions, where the link generation is also done using `path_helper` functions. Making `filter_inputs_for` public would make sense as well, but the primary API should include the `<form>`, in my perspective. Options to the `form_for` could be passed as usual, and additional markup inside the form could be passed as `@inner_block` (or soon, as a slot)?

I'm not sure about the `@inner_block`. Where would you render it? At the end? At the beginning?

The `filter_form` function wouldn't do more than passing the Phoenix form parameters to the Phoenix form function and the Flop.Phoenix parameters to the `filter_inputs_for` function. I'm not sure how useful that can be in the end. Using the `filter_inputs_for` function directly within the inner block allows you to add a `fieldset`, add multiple `fieldsets` calling the `filter_inputs_for` with for different fields each time, displaying or not displaying a submit button wrapped in the markup you need etc.

The form markup is just not as simple as a table or pagination links. The table markup is more or less the same for every table you would ever render, and there's a limited number of layouts for pagination links, all of which can be accomplished with a little bit of CSS without touching the markup.

For the form, it isn't that simple. There are fieldsets, legends, field wrappers, submit buttons, labels, horizontal layouts, vertical layouts, visually hidden labels, grouped inputs with a single label, grouped fields incl. labels in the same row, inline labels for checkboxes and radio buttons, field add-ons, help texts, error texts, input icons etc. You can try to accommodate all of this in a single function with all kinds of options, but that will be hard to work with, and you'll probably still end up with some limitation that requires you to write your own abstractions. And that's why I'd like to handle this differently from the pagination and table components and think from the inside out, as I already mentioned, starting with just the input helper, and gradually add higher level abstractions as long as it makes sense and the API is still usable.

One thing I forgot is another helper to render the hidden inputs for the other Flop parameters (pagination and ordering). Something like:

```elixir
<Flop.Phoenix.filter_hidden_inputs_for meta={@meta} additional_params={[anything: "else"]} />
```

> Would you like us to provide a proof-of-concept implementation as a starting point for further discussion? I'd see if I can arrange for some time to work on this.

If you want to have a go, go ahead, any help is appreciated!

**Comment by woylie** [comment]
Hey @maltoe,

 I implemented `Phoenix.HTML.FormData` for `Flop.Meta` today: https://github.com/woylie/flop_phoenix/pull/66

Some minor things left to do there, but you can already try it out. Short example:

```elixir
<.form let={f} for={@meta} phx-submit="update-filter" phx-change="update-filter">
    <!-- temporary replacement for hidden_inputs_for/1 - change has already been merged to Phoenix.HTML -->
    <%= Flop.Phoenix.filter_hidden_inputs_for(f) %>

    <%= for ff <- inputs_for(f, :filters, default: [%Flop.Filter{field: :name}]) do %>
      <%= hidden_inputs_for ff %>
      <%= label ff, :value, humanize(input_value(ff, :field)) %>
      <%= text_input ff, :value %>
    <% end %>
  </.form>
```

The version above with `inputs_for` probably works better if you want to build a dynamic filter form where you can add and remove filters. For fixed form layouts, we'll still need that `filter_input` function.

**Comment by maltoe** [comment]
> I implemented Phoenix.HTML.FormData for Flop.Meta today

Nice 👍  I'll take a look later today, if I find the time

> The version above with inputs_for probably works better if you want to build a dynamic filter form where you can add and remove filters. For fixed form layouts, we'll still need that filter_input function.

Yes. As said, IMO these don't have that much in common anymore. I'd pursue the dynamic filter form path first as it should give us more value for now.

With regards to "manual" filter form, I think array slots could be a way to go, see example here: https://hexdocs.pm/phoenix_live_view/Phoenix.Component.html#module-named-slots-with-attributes

**Comment by woylie** [comment]
Hey @maltoe,

I added a couple of components for filtering. Please have a look at let me know whether that works for you.

**Comment by maltoe** [comment]
hey @woylie 

thanks for doing this 💜  Now I feel like I kind of pushed you into this work without contributing much. But at the same time, maybe that's exactly what was needed :)

Left a few comments on the last PR https://github.com/woylie/flop_phoenix/pull/73#pullrequestreview-797015431 - otherwise this looks extremely nice and already covers our use-case.

**Comment by woylie** [comment]
Hey @maltoe,

yes, thanks for bringing this up and for all the feedback. One thing missing is better error handling. Now that you can pass the meta to a form, it would make sense to add an `errors` field to the meta struct and have the `validate_and_run` function return meta with errors instead of a changeset in the error case. What do you think? How do you deal with validation errors at the moment?

**Comment by maltoe** [comment]
> How do you deal with validation errors at the moment?

We don't, we fail on changeset result of `validate_and_run` and return a HTTP 400. Or so, maybe we show a generic error instead, not sure atm. 

> have the validate_and_run function return meta with errors instead of a changeset in the error case

That sounds exactly like what is needed.

**Comment by woylie** [comment]
Released in 0.12.0.
